include/algorithms/synthesis/dd_synthesis.hpp

```diff
@@ -70,9 +70,9 @@ namespace syrec {
         auto unifyPath(dd::mEdge src, dd::mEdge const& current, TruthTable::Cube::Set const& p1SigVec, TruthTable::Cube::Set const& p2SigVec, bool const& changePaths, std::unique_ptr<dd::Package<>>& dd) -> dd::mEdge;
         auto shiftingPaths(dd::mEdge const& src, dd::mEdge const& current, std::unique_ptr<dd::Package<>>& dd) -> dd::mEdge;
 
-        auto decoder(TruthTable::CubeMap const& codewords) -> void;
+        auto decoder(TruthTable::CubeMultiMap const& codewords) -> void;
```

Comment:
`multimap`s are comparatively ineffective. In order to maintain the performance for the case that a multi map is not needed, the best option would probably be to introduce a template parameter for that type.
Or, alternatively, inspired by the argumentation above, there should be a second decoder function that handles the new decoding.

Reply:
I think using the templates is a better option. However, is it wrong to define a function template in a cpp file and declare in the hpp file? 
I tried doing this. When I run on debug and release mode, everything seems to work fine. However, I get warnings like the "parameters are not used" or some variables are unused. They are actually being used (or else the the debug and release modes would have resulted in errors). Any idea as to what is happening?

Reply:
Defining templated functions in cpp files requires extra care. The following SO answer provides some useful context for that: https://stackoverflow.com/a/115735
Essentially you have to make sure that all possible templates that you want are properly instantiated.

I am not yet sure whether going with the template functions is really the best way to go about this. The decoder for both scenarios is quite different.

Reply:
true. I was thinking more about the encoding step.